CASSGO-97 Protocol version negotiation doesn't work if server replies with stream id different than 0

worryg0d opened the pull request:
Protocol version negotiation doesn't work if server replies with stream id different than 0

Previously, protocol negotiation didn't work properly when C* was responding with stream id different from 0.
 
This patch changes the way protocol negotiation works. Instead of parsing a supported protocol version from C* error response, the driver tries to connect with each supported protocol starting from the latest.

Patch by Bohdan Siryk; Reviewed by <> for CASSGO-98

worryg0d committed:
Protocol version negotiation doesn't work if server replies with stream id different than 0

Previously, protocol negotiation didn't work properly when C* was responding with stream id different from 0.

This patch changes the way protocol negotiation works. Instead of parsing a supported protocol version from C* error response, the driver tries to connect with each supported protocol starting from the latest.

Patch by Bohdan Siryk; Reviewed by João Reis for CASSGO-98

worryg0d commented:
WIP: currently working on tests for the whole protocol negotiation logic....

worryg0d commented:
Ready for review

joao-r-reis commented on control.go:
```diff
@@ -262,26 +217,27 @@ func (c *controlConn) discoverProtocol(hosts []*HostInfo) (int, error) {
-		}
+			if err == nil {
+				c.session.logger.Debug("Discovered protocol version using host.",
+					NewLogFieldInt("protocol_version", connCfg.ProtoVersion), NewLogFieldIP("host_addr", host.ConnectAddress()))
```
any particular reason you removed the host_id log field from both calls? Is it because it's not populated?

worryg0d replied:
Yes, it is not populated, so there is really no useful information

joao-r-reis commented on protocol_negotiation_test.go:
```diff
@@ -0,0 +1,179 @@
+
+	// if a client uses an unsupported protocol version, we respond with an error
+	if !r.supportsProtocol(reqFrame.header.version) {
+		respFrame.writeHeader(0, opError, reqFrame.header.stream)
```
can we have a test case that replies with the stream id of the request in addition to this one? This behavior is not defined in the spec and I think some C* versions might behave differently

worryg0d replied:
Could you please clarify what you exactly mean? The test server already replies with the stream id of the request

joao-r-reis replied:
Oh I'm sorry I thought it was replying with stream id 0, in that case then can you add a test case where the response stream id is 0 and the request stream id is not 0?

worryg0d replied:
Added test cases for 0 stream id

joao-r-reis commented on control.go:
```diff
@@ -262,26 +217,27 @@ func (c *controlConn) discoverProtocol(hosts []*HostInfo) (int, error) {
-			c.session.logger.Debug("Discovered protocol version using host after parsing protocol error.",
-				NewLogFieldInt("protocol_version", proto), NewLogFieldIP("host_addr", host.ConnectAddress()), NewLogFieldString("host_id", host.HostID()))
-			return proto, nil
+			c.session.logger.Debug("Failed to connect to the host using protocol version.",
```
We should only attempt to reconnect to the same host with a lower protocol version if the error is assumed to be related to unsupported protocol version: 
- t's an error to the first request (OPTIONS)
- the error type is PROTOCOL_ERROR or SERVER_ERROR - SERVER_ERROR is for old C* versions that reported this error as SERVER_ERROR instead of PROTOCOL_ERROR

I propose we use a custom internal error type (unsupportedProtocolVersionErr for example) just for this case that we can return inside the dial method and we test for it here in discoverProtocol(). Reference code [in the java driver here](https://github.com/apache/cassandra-java-driver/blob/62eade21bfeb16a12ce71013fbaebf1c19b5ae96/core/src/main/java/com/datastax/oss/driver/internal/core/channel/ChannelFactory.java#L238-L259).

[Technically the java driver does check for the error string](https://github.com/apache/cassandra-java-driver/blob/62eade21bfeb16a12ce71013fbaebf1c19b5ae96/core/src/main/java/com/datastax/oss/driver/internal/core/channel/ProtocolInitHandler.java#L327-L334) which I didn't know but I still believe we shouldn't do that, checking the above 2 conditions is enough.

worryg0d replied:
That makes sense, we don't really have to re-try each time if the error is not protocol-related

joao-r-reis commented on protocol_negotiation_test.go:
```diff
@@ -0,0 +1,179 @@
+//go:build all || unit
+// +build all unit
+
+package gocql
```
add license header, we should add a check for this in CI at some point

worryg0d replied:
Good catch!

joao-r-reis commented on protocol_negotiation_test.go:
```diff
@@ -0,0 +1,179 @@
+	return nil
+}
+
+func TestProtocolNegotiation(t *testing.T) {
```
Very good job with this test, I feel like it was a weak spot in our test suite and now it's much better. Can you add `-v` to the `go test` command in the GH workflow step that runs the unit tests so we can double check which tests are running?

worryg0d replied:
Thanks. I'll add verbose for unit tests to ci

joao-r-reis reviewed (commented):
Looking good

worryg0d commented:
Thanks, @joao-r-reis, for your feedback. I'm going to address this next week

worryg0d commented:
Hi @joao-r-reis

Sorry for the late reply — I’ve been a bit tied up with work lately.

I’ve now addressed your comments, so please take a look whenever you have a moment.

Thanks.

joao-r-reis commented on conn.go:
```diff
@@ -659,6 +700,8 @@ func (c *Conn) processFrame(ctx context.Context, r io.Reader) error {
 		return err
 	}
 
+	c.logger.Debug("Received frame header", NewLogFieldString("header", head.String()))
```
Is this `processFrame` function called on every frame or just handshake frames? Running this log call on every frame post handshake seems a bit unnecessary no?

worryg0d replied:
Nice catch. I needed this when I was implementing it, but it is not meant to be left here. Will remove it

worryg0d replied:
Removed

joao-r-reis commented on logger.go:
```diff
@@ -162,7 +162,8 @@ func (l *defaultLogger) Debug(msg string, fields ...LogField) {
 	writeLogMsg(&buf, prefix, msg, fields)
-	log.Println(buf.String())
+	// During testing it allows seeing where exactly logging methods are called instead of this method
+	log.Output(3, buf.String())
```
This is used in production code as well, I'd prefer to keep this behavior unchanged for users

worryg0d replied:
Oh, I assumed that this for testing purpose only... Well if this is the case I'm going to revert this change. But it would actually useful to see the exact caller...

joao-r-reis replied:
Yeah I agree that the caller info is very useful and I'd +1 that if it was only used in tests

worryg0d replied:
Reverted this change

joao-r-reis reviewed (commented):
Looks good, left 2 comments only

worryg0d commented:
Addressed review comments, squashed commits, and added @joao-r-reis as a reviewer to the commit message

joao-r-reis reviewed (approved):
👍

worryg0d commented:
Thanks for merging it
